chained/water_impl.go

```diff
@@ -39,40 +38,44 @@ func newWaterImpl(addr string, pc *config.ProxyConfig, reportDialCore reportDial
 		}
 	}
 
+	ctx := context.Background()
 	wasmAvailableAt := ptSetting(pc, "water_available_at")
+	transport := ptSetting(pc, "water_transport")
 	if wasm == nil && wasmAvailableAt != "" {
-		urls := strings.Split(wasmAvailableAt, ",")
-		b := new(bytes.Buffer)
-		cli := httpClient
-		if cli == nil {
-			downloadTimeout := ptSetting(pc, "water_download_timeout")
-			timeout := 10 * time.Minute
-			if downloadTimeout != "" {
-				var err error
-				timeout, err = time.ParseDuration(downloadTimeout)
-				if err != nil {
-					return nil, fmt.Errorf("failed to parse download timeout: %w", err)
-				}
-			}
-
-			cli = proxied.ChainedThenDirectThenFrontedClient(timeout, "")
+		waterDir := filepath.Join(configDir, "water")
+		vc, err := NewVersionControl(waterDir)
+		if err != nil {
+			return nil, log.Errorf("failed to create version control: %w", err.Error())
 		}
-		d, err := NewWASMDownloader(urls, cli)
+
+		r, err := vc.GetWASM(ctx, transport, strings.Split(wasmAvailableAt, ","))
 		if err != nil {
-			return nil, log.Errorf("failed to create wasm downloader: %s", err.Error())
+			return nil, log.Errorf("failed to get wasm: %w", err)
 		}
-		if err = d.DownloadWASM(context.Background(), b); err != nil {
-			return nil, log.Errorf("failed to download wasm: %s", err.Error())
+		defer r.Close()
+
+		b, err := io.ReadAll(r)
+		if err != nil {
+			return nil, log.Errorf("failed to read wasm: %w", err)
 		}
-		wasm = b.Bytes()
+
+		if len(b) == 0 {
+			return nil, log.Errorf("received empty wasm")
+		}
+
+		if err = vc.Commit(transport); err != nil {
```

Comment:
Couldn't `GetWASM` do this?  Any time a WASM is fetched, we know it's being used.